src/audio_device.c

```diff
@@ -46,125 +42,174 @@
-
-    uint32_t source, buffers[OPENAL_BUFS]; /* Playback source/buffers */
-    uint32_t ref_count;
-    int32_t selection;
-    bool enable_VAD;
+    bool active;
     bool muted;
-    pthread_mutex_t mutex[1];
-    uint32_t sample_rate;
-    uint32_t frame_duration;
-    int32_t sound_mode;
-#ifdef AUDIO
-    float VAD_treshold;                    /* 40 is usually recommended value */
-#endif
+
+    FrameInfo frame_info;
+
+    // used only by input devices:
+    DataHandleCallback cb;
+    void *cb_data;
+    float VAD_threshold;
+    uint32_t VAD_samples_remaining;
+
+    // used only by output devices:
+    uint32_t source, buffers[OPENAL_BUFS];
+    bool source_open;
 } Device;
 
-const char *ddevice_names[2];              /* Default device */
-const char *devices_names[2][MAX_DEVICES]; /* Container of available devices */
-static int size[2];                        /* Size of above containers */
-Device *running[2][MAX_DEVICES] = {{NULL}};     /* Running devices */
-uint32_t primary_device[2];          /* Primary device */
+typedef struct AudioState {
+    ALCdevice *al_device[2];
 
-#ifdef AUDIO
-static ToxAV *av = NULL;
-#endif /* AUDIO */
+    Device devices[2][MAX_DEVICES];
+    uint32_t num_devices[2];
+
+    FrameInfo capture_frame_info;
+    float input_volume;
+
+    // mutexes to prevent changes to input resp. output devices and al_devices
+    // during poll_input iterations resp. calls to write_out;
+    // mutex[input] also used to lock input_volume which poll_input writes to.
+    pthread_mutex_t mutex[2];
 
-/* q_mutex */
-#define lock pthread_mutex_lock(&mutex)
-#define unlock pthread_mutex_unlock(&mutex)
-pthread_mutex_t mutex;
+    // TODO: unused
+    const char *default_al_device_name[2];              /* Default devices */
+
+    const char *al_device_names[2][MAX_OPENAL_DEVICES]; /* Available devices */
+    uint32_t num_al_devices[2];
+    char *current_al_device_name[2];
+} AudioState;
+
+AudioState *audio_state;
+
+void lock(DeviceType type)
+{
+    pthread_mutex_lock(&audio_state->mutex[type]);
+}
+
+void unlock(DeviceType type)
+{
+    pthread_mutex_unlock(&audio_state->mutex[type]);
+}
 
 
 bool thread_running = true,
      thread_paused = true;               /* Thread control */
 
-void *thread_poll(void *);
-/* Meet devices */
 #ifdef AUDIO
-DeviceError init_devices(ToxAV *av_)
-#else
+static void *poll_input(void *);
+#endif
+
+static uint32_t sound_mode(bool stereo)
+{
+    return (stereo ? AL_FORMAT_STEREO16 : AL_FORMAT_MONO16);
+}
+
+static uint32_t sample_size(bool stereo)
+{
+    return (stereo ? 4 : 2);
+}
+
 DeviceError init_devices(void)
-#endif /* AUDIO */
 {
-    get_devices_names();
+    audio_state = calloc(1, sizeof(AudioState));
 
-    // Start poll thread
-    if (pthread_mutex_init(&mutex, NULL) != 0) {
+    if (audio_state == NULL) {
         return de_InternalError;
     }
 
+    get_al_device_names();
+
+    for (DeviceType type = input; type <= output; type++) {
```

Comment:
++type (but I guess I won't mention these any further. Maybe at some point we can throw tokstyle at this code.)